app/services/fact_importer.rb

```diff
@@ -91,16 +93,18 @@ def create_fact_name(fact_names, name, fact_value)
   end
 
   def update_facts
-    facts_to_update = []
-    db_facts.each { |name, fv| facts_to_update << [facts[name], fv] if fv.value != facts[name] }
-
-    @counters[:updated] = facts_to_update.size
-    return logger.debug("No facts update required for #{host}") if facts_to_update.empty?
-
-    logger.debug("Merging facts for '#{host}': updated #{@counters[:updated]} facts")
-
-    facts_to_update.each do |new_value, fv|
-      fv.update_attribute(:value, new_value)
+    @counters[:updated] = 0
+    db_facts_as_relation.find_each do |record|
+      new_value = facts[record.name]
+      if record.value != new_value
+        record.update_attribute(:value, new_value)
```

Comment:
I think the only thing more that I think can be optimized is if we can somehow update all the facts in one query, but I'm not sure it's possible in this case.
Note that by skipping the callbacks we don't get the updated_at field updated, but it can be updated in one query using the `update_all` method.